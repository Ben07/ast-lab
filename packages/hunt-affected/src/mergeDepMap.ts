import path from 'path';
import enhancedResolve from 'enhanced-resolve';
import _debug from 'debug';
import denodeify from './denodeify';
import { DependencyMap } from 'ast-lab-types';
import { Options } from './types';
import fileDepMap from './fileDepMap';
import appendEntries from './appendEntries';

const debug = _debug('hunt-affected:merge');
const core = new Set(require('module').builtinModules);
/**
 * Extract the dependency map of a list of source files.
 * @param sources A list of absolute source file paths
 * @param opts
 */
export default async function mergeDepMap(sources: string[], opts: Options = {}) {
  debug('===');
  const srcSet = new Set(sources);
  const depMap = new Map() as DependencyMap;
  const resolver = opts.resolver || denodeify(enhancedResolve);

  const fileHandlers = [] as Array<Promise<any>>;
  sources.forEach((src) => {
    fileHandlers.push((async () => {
      const fileDependencyMap = await fileDepMap(src, opts);
      const depMapHandlers = [] as Array<Promise<any>>;
      fileDependencyMap.forEach((memberDeps, modRelativePath) => {
        debug(`${src} > ${modRelativePath} start`);
        depMapHandlers.push((async () => {
          const baseDir = path.dirname(src);
          let modPath = '';
          if (path.isAbsolute(modRelativePath) || core.has(modRelativePath)) {
            modPath = modRelativePath;
          } else {
            try {
              modPath = await resolver(baseDir, modRelativePath) || '';
            } catch (e) {
              // do nothing
            }
            if (!modPath) {
              console.warn(`Could not solve ${modRelativePath} from ${src}.`);
              modPath = /\.\w+$/.test(modRelativePath) ? 
                path.resolve(baseDir, modRelativePath) :
                modRelativePath;
            } else if(/node_modules\//.test(modPath)) {
              modPath = modRelativePath;
            }
          }
          if (depMap.get(modPath)) {
            debug(`${src} > ${modPath} append entries: ${JSON.stringify(Array.from(memberDeps.entries()))}`);
            memberDeps.forEach((entries, member) => {
              appendEntries(depMap, modPath, member, entries);
            });
          } else if (srcSet.has(modPath)) {
            debug(`${src} > ${modPath} add new: ${JSON.stringify(Array.from(memberDeps.entries()))}`);
            depMap.set(modPath, memberDeps);
          } else {
            debug(`${src} > ${modPath} depMap not added!`);
          }
        })());
      });
      await Promise.all(depMapHandlers);
    })());
  });
  await Promise.all(fileHandlers);
  return depMap;
}