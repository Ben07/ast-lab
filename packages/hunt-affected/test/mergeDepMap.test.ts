import fs from 'fs';
import path from 'path';
import _debug from 'debug';
import mergeDepMap from '../src/mergeDepMap';
import { relativeDepMap } from './util';

const debug = _debug('hunt-affected:test');
const tests = [
  'exports',
  'imports'
]
describe('mergeDepMap()', () => {
  tests.forEach(_dir => {
    const dir = path.resolve(__dirname, '__fixtures__', _dir);
    const sources = fs.readdirSync(dir);
    test(_dir, async () => {
      debug(_dir);
      const res = await mergeDepMap(
        sources.map(src => path.resolve(dir, src)),
        {
          parserOptions: {
            plugins: ['dynamicImport']
          }
        }
      );
      expect(relativeDepMap(res)).toMatchSnapshot();
    });
  });
});
