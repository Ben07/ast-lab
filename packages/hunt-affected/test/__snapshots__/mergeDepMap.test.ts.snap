// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeDepMap() exports 1`] = `
Map {
  "exports/exportDefault.js" => Map {
    "default" => Array [],
  },
  "exports/exportNamed.js" => Map {
    "a" => Array [
      Object {
        "name": "default",
        "source": "exports/exportDefault.js",
      },
      Object {
        "name": "a",
        "source": "exports/exportNamedFrom.js",
      },
    ],
    "readFile" => Array [],
    "*" => Array [
      Object {
        "name": "*",
        "source": "exports/exportAllFrom.js",
      },
    ],
  },
  "exports/exportNamedFrom.js" => Map {
    "a" => Array [],
  },
}
`;

exports[`mergeDepMap() imports 1`] = `
Map {
  "imports/dynamicImport-asyncawait.js" => Map {
    "func" => Array [],
  },
  "imports/dynamicImport-then.js" => Map {
    "func" => Array [],
  },
  "imports/importAlias.js" => Map {
    "c" => Array [],
    "func" => Array [],
  },
  "imports/importAll.js" => Map {
    "a" => Array [],
    "func" => Array [],
    "default" => Array [
      Object {
        "name": "func",
        "source": "imports/dynamicImport-asyncawait.js",
      },
    ],
  },
  "imports/importDefault.js" => Map {
    "a" => Array [],
    "func" => Array [],
    "default" => Array [
      Object {
        "name": "func",
        "source": "imports/dynamicImport-then.js",
      },
    ],
    "*" => Array [
      Object {
        "name": "a",
        "source": "imports/importAll.js",
      },
      Object {
        "name": "func",
        "source": "imports/importAll.js",
      },
    ],
  },
  "imports/importNamed.js" => Map {
    "resolveFactory" => Array [
      Object {
        "name": "c",
        "source": "imports/importAlias.js",
      },
      Object {
        "name": "func",
        "source": "imports/importAlias.js",
      },
    ],
    "func" => Array [],
    "default" => Array [],
  },
}
`;
